lib/spack/spack/test/abi_splicing.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+""" Test ABI-based splicing of dependencies """
+
+from typing import List
+
+import pytest
+
+import spack.config
+import spack.package_base
+import spack.paths
+import spack.repo
+import spack.solver.asp
+import spack.spec
+from spack.installer import PackageInstaller
+from spack.spec import Spec
+
+
+class CacheManager:
+    def __init__(self, specs: List[str]) -> None:
+        self.req_specs = specs
+        self.concr_specs: List[Spec]
+        self.concr_specs = []
+
+    def __enter__(self):
+        self.concr_specs = [Spec(s).concretized() for s in self.req_specs]
+        for s in self.concr_specs:
+            PackageInstaller([s.package], fake=True, explicit=True).install()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for s in self.concr_specs:
+            s.package.do_uninstall()
+
+
+# MacOS and Windows only work if you pass this function pointer rather than a
+# closure
+def _mock_has_runtime_dependencies(_x):
+    return True
+
+
+def _make_specs_non_buildable(specs: List[str]):
+    output_config = {}
+    for spec in specs:
+        output_config[spec] = {"buildable": False}
+    return output_config
+
+
+@pytest.fixture
+def splicing_setup(mutable_database, mock_packages, monkeypatch):
+    spack.config.set("concretizer:reuse", True)
+    monkeypatch.setattr(
+        spack.solver.asp, "_has_runtime_dependencies", _mock_has_runtime_dependencies
+    )
+
+
+def _enable_splicing():
+    spack.config.set("concretizer:splice", {"automatic": True})
+
+
+def test_simple_reuse(splicing_setup):
+    with CacheManager(["splice-z@1.0.0+compat"]):
+        spack.config.set("packages", _make_specs_non_buildable(["splice-z"]))
+        Spec("splice-z").concretized()
+    assert True
```

Comment:
It looks like most of these tests are missing the asserts to actually check what they're testing. Are they all relying on the satisfaction check at the end of concretization for their assertion? If so, we should duplicate the check here (so that changes to that code that break the assumptions here are caught by these tests).

Reply:
Yes, most of the tests have the general structure of: Catch unsat without splicing, succeed with satisfaction with splicing. Replicating the satisfaction should not be too hard.